records/2021-08-16.scalability.move-api-to-lambda.md

```diff
@@ -0,0 +1,64 @@
+# Move the Notify API from Kubernetes to AWS Lambda
+
+Date: 2021-08-16
+
+## Status
+
+**PROPOSED**
+
+## Context
+
+The SRE team is encouraging us to move the Notify API from Kubernetes to AWS Lambda. AWS Lambda functions allow code to be run in a highly scalable way - we can have hundreds of lambda functions running simulateously, and this auto-scales to meet the number of requests being made to the API. The benefits of doing this are thought to be the following:
+
+- less complexity for the Notify team to manage, compared to Kubernetes
+- more scalable API
+- less costly compared to kubernetes + Amazon EKS
+
+At the time when Notify was being deployed by CDS, AWS Lambda had a smaller maximum size for a lambda function. The notification-api codebase was over this size limit, so it was deployed a different way.
+
+Since then, AWS has added support for Docker images and increased the size limit for these to 10 GB. This means it is now possible to deploy the Notify API as a Lambda function.
+
+CDS has now used lambda functions to run the [Covid Alert metrics collection and aggregation](https://github.com/cds-snc/covid-alert-metrics-terraform/tree/main/aws). Lambda functions helped that part of the service to scale well.
+
+## Proposed architecture
+
+AWS Lambda (with some amount of provisioned concurrency) + API Gateway
+
+## Additional considerations
+
+AWS Lambda functions exhibit "cold-start" behaviour, where they can take more than 1 second to respond if they [have not been invoked recently](https://aws.amazon.com/blogs/compute/operating-lambda-performance-optimization-part-1/#:~:text=The%20duration%20of%20a%20cold,test%20functions%20than%20production%20workloads). If we do not mitigate this, the admin site will appear unresponsive at times. AWS has a feature to mitigate this, called [provisioned concurrency](https://aws.amazon.com/blogs/compute/new-for-aws-lambda-predictable-start-up-times-with-provisioned-concurrency/).
+
+Another consideration will be the effects on other parts of Notify. If the API can scale much better than in the past, will some other parts of the system become overloaded? This may not be an issue because we queue our upcoming jobs with Celery, but we should do significant load testing before running lambda functions in production.
+
+Lambdas may make local testing a bit more difficult by forcing us to either wire up some docker containers using aws-lambda-rie or use AWS SAM to orchestrate the work. With Kubernetes, testing locally is easier since docker desktop can be used to bring up the test system.
+
+## Load testing results
+
+We ran a load test where we simulated 5,500 requests per minute to the api using locust, to see if the new lambda deployment could meet existing peak loads. The these requests were made to the `/v2/notifications/email` endpoint and were split into 2 services to make the test more realistic. Here are the reports:
+
```

Comment:
It might be worth describing the pacing time and how bursty the tests were and add how unusual it is from regular API loads in production (hence the failure is "acceptable" because the traffic is out of normal bounds), but yet shows that lambdas can withstand more compared to our current EKS/machines/pods configuration.

Reply:
I'd also be curious what the load test says without a provisioned lambda? What kind of times do you get with all cold starts? (Or do you expect your traffic to be on the order of 1 request in the time it takes the lambda to shutdown?)

Reply:
> I'd also be curious what the load test says without a provisioned lambda? 

We determined from the AWS logs that only about 10% of requests were handled by the provisioned lambdas, and the other 90% were handled by non-provisioned lambdas. So the response time stats were dominated by the responses of non-provisioned lambdas. Therefore I would expect these stats wouldn't change by much if we hadn't provisioned any lambdas. 

This chart is a bit hard to read but it shows the approx. 90%/10% split:
![image (2)](https://user-images.githubusercontent.com/5498428/130139506-722504eb-90e8-4435-b4e1-fe96aaad3236.png)